linter/lints/crl/lints.go

```diff
@@ -369,3 +372,135 @@ func hasMozReasonCodes(crl *crl_x509.RevocationList) *lint.LintResult {
+		Status:  lint.Error,
+		Details: "Failed to re-parse tbsCertList during linting",
+	}
+	// Read tbsCertList.
+	var tbs cryptobyte.String
+	input.ReadASN1(&tbs, cryptobyte_asn1.SEQUENCE)
+	// Skip (optional) version.
+	if !tbs.SkipOptionalASN1(cryptobyte_asn1.INTEGER) {
+		return &parseFail
+	}
+	// Skip signature.
+	if !tbs.SkipASN1(cryptobyte_asn1.SEQUENCE) {
+		return &parseFail
+	}
+	// Skip issuer.
+	if !tbs.SkipASN1(cryptobyte_asn1.SEQUENCE) {
+		return &parseFail
+	}
+	// Read thisUpdate.
+	var thisUpdate cryptobyte.String
+	var thisUpdateTag cryptobyte_asn1.Tag
+	if !tbs.ReadAnyASN1Element(&thisUpdate, &thisUpdateTag) {
+		return &parseFail
+	}
+	// Lint thisUpdate.
+	err := lintTimestamp(&thisUpdate, thisUpdateTag)
+	if err != nil {
+		parseFail.Details = err.Error()
+		return &parseFail
+	}
+	// Peek (optional) nextUpdate.
+	if tbs.PeekASN1Tag(cryptobyte_asn1.UTCTime) || tbs.PeekASN1Tag(cryptobyte_asn1.GeneralizedTime) {
+		// Read nextUpdate.
+		var nextUpdate cryptobyte.String
+		var nextUpdateTag cryptobyte_asn1.Tag
+		if !tbs.ReadAnyASN1Element(&nextUpdate, &nextUpdateTag) {
+			return &parseFail
+		}
+		// Lint nextUpdate.
+		err = lintTimestamp(&nextUpdate, nextUpdateTag)
+		if err != nil {
+			parseFail.Details = err.Error()
+			return &parseFail
+		}
+	}
+	// Read sequence of revokedCertificate.
+	var revokedSeq cryptobyte.String
+	tbs.ReadASN1(&revokedSeq, cryptobyte_asn1.SEQUENCE)
+	for !revokedSeq.Empty() {
+		// Read revokedCertificate.
+		var certSeq cryptobyte.String
+		revokedSeq.ReadASN1Element(&certSeq, cryptobyte_asn1.SEQUENCE)
+		certSeq.ReadASN1(&certSeq, cryptobyte_asn1.SEQUENCE)
+
+		// Skip userCertificate (serial number).
+		certSeq.SkipASN1(cryptobyte_asn1.INTEGER)
+
+		// Read revocationDate.
+		var revocationDate cryptobyte.String
+		var revocationDateTag cryptobyte_asn1.Tag
+		if !certSeq.ReadAnyASN1Element(&revocationDate, &revocationDateTag) {
+			return &parseFail
+		}
+
+		// Lint revocationDate.
+		err = lintTimestamp(&revocationDate, revocationDateTag)
+		if err != nil {
+			parseFail.Details = err.Error()
+			return &parseFail
+		}
+	}
+	return &lint.LintResult{Status: lint.Pass}
+}
+
+func lintTimestamp(der *cryptobyte.String, tag cryptobyte_asn1.Tag) error {
+	utcTimeFormat := "YYMMDDHHMMSSZ"
+	generalizedTimeFormat := "YYYYMMDDHHMMSSZ"
+	// Preserve the original timestamp for length checking.
+	derBytes := *der
+	var tsBytes cryptobyte.String
+	derBytes.ReadASN1(&tsBytes, tag)
+	tsString := string(tsBytes)
+	var t time.Time
+	switch tag {
+	case cryptobyte_asn1.UTCTime:
+		// Verify that the timestamp is properly formatted.
+		if len(tsString) != len(utcTimeFormat) {
+			return fmt.Errorf("x509: timestamps encoded using UTCTime MUST be specified in the format %q", utcTimeFormat)
+		}
+		// Verify that the timestamp is prior to the year 2050.
+		der.ReadASN1UTCTime(&t)
+		if t.Year() > 2049 {
+			return errors.New("x509: timestamps after 2049 MUST be encoded using GeneralizedTime")
+		}
```

Comment:
This is nice but ultimately useless, I believe. The `ReadASN1UTCTime` function is incapable of returning a `time.Time` later than 2050. If this conditional were ever met it would indicate a bug in that code, not an incorrect encoding in the CRL itself.

Reply:
You're absolutely correct concerning the `ReadASN1UTCTime`. My reasoning was to err on the side of caution and correctness given that we don't have control over that library. I didn't think it would be correct for the guarantees of the lint to be entirely dependent on a method that in other regards (missing seconds) fails to meet the time format specified in RFC 5280.

Reply:
Yeah to be clear I'm not saying that you _should_ remove it. It's just that... suppose the date successfully parses as a UTCTime. Therefore we _know_ that it only had two digits representing the year. Therefore we know, _regardless_ of what the library call tells us, that it cannot possibly represent a date later than 2049, _by definition_.

So even if the library started mis-interpreting UTCTimes as being after 2049, those timestamps would _actually be correctly encoded_. There might be some other error, like accidentally setting 1951 for the nextUpdate and having a validity period of negative ninety nine years, but the timestamp itself in isolation would be a valid UTCTime.

Reply:
I don't believe a UTCTime encoded timestamp of >2050 is valid according to RFC 5280 section [4.1.2.5.1](https://www.rfc-editor.org/rfc/rfc5280.html#section-4.1.2.5.1):

```
Conforming systems MUST interpret the year field (YY) as follows:

      Where YY is greater than or equal to 50, the year SHALL be
      interpreted as 19YY; and

      Where YY is less than 50, the year SHALL be interpreted as 20YY.
```

Reply:
Oh, gosh, yeah I get what you're saying now. The key phrase here is:

> Therefore we know, regardless of what the library call tells us, that it cannot possibly represent a date later than 2049, by definition.

Yeah, absolutely. Given this, what purpose will this serve other than as a canary that the library we depend on is broken? Is that sufficient reason to leave it in?

Reply:
Yeah exactly. I like having it due to the symmetry. But I think overall it's better to just replace it with a comment to the effect of this conversation.

Reply:
I've decided that I'd like to know if this happens. We definitely have code that loads CRLs and depends on this being parsed correctly. I've made the error more specific to the actual underlying problem. Let me know what you think.

Reply:
Sweet, I think that's a totally reasonable call. Having a specialized error message for it makes sense. Just make sure that all of the error messages are things that make sense as lint results, as opposed to the sorts of things we normally just toss into logs.